pkg/propertyprovider/aks/trackers/trackers.go

```diff
@@ -0,0 +1,570 @@
+	nt.untrackSKU(nodeName)
+
+	// Re-calculate costs.
+	//
+	// Note that it would be very rare for the informer to receive a node deletion event before
+	// having a chance to track it first.
+	nt.calculateCosts()
+}
+
+// NodeCount returns the node count stat that a node tracker tracks.
+func (nt *NodeTracker) NodeCount() int {
+	nt.mu.RLock()
+	defer nt.mu.RUnlock()
+
+	// capacityByNode and allocatableByNode should always have the same length.
+	return len(nt.allocatableByNode)
+}
+
+// TotalCapacityFor returns the total capacity of a specific resource that the node
+// tracker tracks.
+func (nt *NodeTracker) TotalCapacityFor(rn corev1.ResourceName) resource.Quantity {
+	nt.mu.RLock()
+	defer nt.mu.RUnlock()
+
+	return nt.totalCapacity[rn]
+}
+
+// TotalAllocatableFor returns the total allocatable capacity of a specific resource that the node
+// tracker tracks.
+func (nt *NodeTracker) TotalAllocatableFor(rn corev1.ResourceName) resource.Quantity {
+	nt.mu.RLock()
+	defer nt.mu.RUnlock()
+
+	return nt.totalAllocatable[rn]
+}
+
+// TotalCapacity returns the total capacity of all resources that the node tracker tracks.
+func (nt *NodeTracker) TotalCapacity() corev1.ResourceList {
+	nt.mu.RLock()
+	defer nt.mu.RUnlock()
+
+	return nt.totalCapacity
+}
+
+// TotalAllocatable returns the total allocatable capacity of all resources that
+// the node tracker tracks.
+func (nt *NodeTracker) TotalAllocatable() corev1.ResourceList {
+	nt.mu.RLock()
+	defer nt.mu.RUnlock()
+
+	return nt.totalAllocatable
+}
+
+// Costs returns the per CPU core and per GB memory costs in the cluster.
+func (nt *NodeTracker) Costs() (perCPUCoreCost, perGBMemoryCost float64, err error) {
+	nt.mu.Lock()
+	defer nt.mu.Unlock()
+	if nt.costLastUpdated.Before(nt.pricingProvider.LastUpdated()) {
+		nt.calculateCosts()
+	}
+	return nt.perCPUCoreHourlyCost, nt.perGBMemoryHourlyCost, nt.costErr
+}
+
+// NodeTracker helps track specific stats about nodes in a Kubernetes cluster, e.g., the sum
+// of its requested resources.
+type PodTracker struct {
+	totalRequested corev1.ResourceList
+
+	requestedByPod map[string]corev1.ResourceList
+
+	// mu is a RWMutex that protects the tracker against concurrent access.
+	mu sync.RWMutex
+}
+
+// NewPodTracker returns a pod tracker.
+func NewPodTracker() *PodTracker {
+	pt := &PodTracker{
+		totalRequested: make(corev1.ResourceList),
+		requestedByPod: make(map[string]corev1.ResourceList),
+	}
+
+	for _, rn := range supportedResourceNames {
+		pt.totalRequested[rn] = resource.Quantity{}
+	}
+
+	return pt
+}
+
+// AddOrUpdate starts tracking a pod or updates the stats about a pod that has been
+// tracked.
+func (pt *PodTracker) AddOrUpdate(pod *corev1.Pod) {
+	pt.mu.Lock()
+	defer pt.mu.Unlock()
+
+	requestsAcrossAllContainers := make(corev1.ResourceList)
+	for _, container := range pod.Spec.Containers {
+		for _, rn := range supportedResourceNames {
+			r := requestsAcrossAllContainers[rn]
+			r.Add(container.Resources.Requests[rn])
```

Comment:
what if request is not set?
what if both request and limit is not set?

Reply:
Per Go's language spec, reading a nil map is the same as reading an empty map, so it will yield a zero resource.Quantity.

Reply:
so , if the request is not set, we will ignore this pod. If i understand it correctly, the cost may be not accurate because of missing this config.

Could you please add a comment to record this behavior?

Reply:
Discussed offline. For reference, the costs are not calc'd using requested capacities; and due to technical limitations at this moment we cannot reliably tell how much resource a pod currently utilizes